draft-ietf-ppm-dap.md

```diff
@@ -1305,50 +1500,97 @@ given batch interval, denoted `batch_interval`, decryption works as follows:
-  batch of an AggregateShareReq. If any report in `X` was added to at least
-  `max_batch_lifetime` previous batches, then the aggregator MUST abort and
-  alert the peer with "batchLifetimeExceeded".
+Next, the Aggregator checks that batch contains a valid number of reports, as
+determined by the query type. If the size check fails, then the Aggregator MUST
+abort with error "invalidBatchSize".
 
-Finally, the aggregator checks that the batch interval defined by the collect
-request satisfies one of the conditions:
+Next, the Aggregator checks that the batch has not been aggregated too many
+times. This is determined by the maximum batch lifetime, `max_batch_lifetime`.
+Unless the query has been issued less than `max_batch_lifetime` times, the
+Aggregator MUST abort with error "batchLifetimeExceeded".
 
-1. The batch interval does not overlap with the batch interval of any prior
-   completed collect requests.
-1. The batch interval, including its start and duration values, match a prior
-   completed collect request.
+Finally, the Aggregator checks that the batch does not contain a report that was
+included in any previous batch. If this batch overlap check fails, then the
+Aggregator MUST abort with error "batchOverlap". For time-interval tasks, it is
+sufficient (but not necessary) to check that the batch interval does not overlap
+with the batch interval of any previous query. If this batch interval check
+fails, then the Aggregator MAY abort with error "batchOverlap".
 
 [[OPEN ISSUE: #195 tracks how we might relax this constraint to allow for more
 collect query flexibility. As of now, this is quite rigid and doesn't give the
 collector much room for mistakes.]]
 
+#### Time-interval Queries {#time-interval-batch-validation}
+
+##### Boundary Check
+
+The batch boundaries are determined by the `time_precision` field of the query
+configuration. For the `batch_interval` included with the query, the Aggregator
+checks that:
+
+* `batch_interval.duration >= time_precision` (this field determines,
+  effectively, the minimum batch duration)
+
+* both `batch_interval.start` and `batch_interval.duration` are divisible by
+  `time_precision`
+
+These measures ensure that Aggregators can efficiently "pre-aggregate" output
+shares recovered during the aggregation sub-protocol.
+
+##### Size Check
+
+The query configuration specifies the minimum batch size, `min_batch_size`. The
+Aggregator checks that `len(X) >= min_batch_size`, where `X` is the set of
+reports in the batch.
+
+#### Fixed-size Queries {#fixed-size-batch-validation}
+
+##### Boundary Check
+
+For fixed-size tasks, the batch boundaries are defined by "batch IDs" selected
+by the Leader. Thus the Aggregator needs to check that the query is associated
+with a known batch ID:
+
+* If the query contained in the CollectReq has type `by-batch-id`, the
+  Leader checks that the provided batch ID corresponds to a batch ID it
+  returned in a previous CollectResp for the task.
```

Comment:
I don't think this requirement is necessary. The Collector should be able to query for batches by ID opportunistically, I think.

Reply:
I don't see a way to enable this without creating lots of corner cases. I think we need two things in order to pursue this:
1. What's the motivation? Is there a reason why the current scheme is not workable for Collectors? What's the advantage of knowing the batch ID in advance, beyond not having to store it?
2. A way to address corner cases like: What does the Collector do if the Leader requests batch 9999, then batch 1? If the Leader is aggregating opportunistically (like it may do for Prio), it might have to wait to starting filling batch 9999 until 1, ..., 9998 have been filled.

This point has already been discussed at some length on the slack. I'm happy to keep discussing, but could we do so in an issue so we have a chance to level set?